great_expectations/core/config_provider.py

```diff
@@ -4,26 +4,45 @@
 import os
 from abc import ABC, abstractmethod
 from collections import OrderedDict
-from typing import Dict, Optional, Type, cast
+from typing import Any, Dict, Optional, Type, cast
 
+from great_expectations.core.config_substitutor import ConfigurationSubstitutor
 from great_expectations.core.yaml_handler import YAMLHandler
 from great_expectations.data_context.types.base import GeCloudConfig
-from great_expectations.data_context.util import (
-    substitute_all_config_variables,
-    substitute_config_variable,
-)
 
 yaml = YAMLHandler()
 
 
 class AbstractConfigurationProvider(ABC):
+    def __init__(self) -> None:
+        self._substitutor = ConfigurationSubstitutor()
+
     @abstractmethod
     def get_values(self) -> Dict[str, str]:
         """
         Retrieve any configuration variables relevant to the provider's environment.
         """
         pass
 
+    def substitute_config(
+        self, config: Any, config_values: Optional[Dict[str, str]] = None
```

Comment:
Do we know anything about the type of `config` here?

Reply:
So there are a few options per the method signature (which was ported over from a util file) - `BaseYamlConfig`, `dict`, `list`, or a catch-all "other".

I'd like to make this more specific but current behavior is essentially `Any`.